pubsub/drivertest/drivertest.go

```diff
@@ -464,6 +465,88 @@ func testBatching(t *testing.T, newHarness HarnessMaker) {
 	}
 }
 
+func testDoubleAck(t *testing.T, newHarness HarnessMaker) {
+	// Set up.
+	ctx := context.Background()
+	h, err := newHarness(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	dt, topicCleanup, err := h.CreateTopic(ctx, t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer topicCleanup()
+	ds, subCleanup, err := h.CreateSubscription(ctx, dt, t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer subCleanup()
+	if ds.AckFunc() != nil {
+		t.Skip("Ack not supported")
+	}
+
+	// Publish 3 messages.
+	for i := 0; i < 3; i++ {
+		err := dt.SendBatch(ctx, []*driver.Message{{Body: []byte(strconv.Itoa(i))}})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Retrieve the messages.
+	var dms []*driver.Message
+	for len(dms) != 3 {
+		curdms, err := ds.ReceiveBatch(ctx, 3)
+		if err != nil {
+			t.Fatal(err)
+		}
+		dms = append(dms, curdms...)
+	}
+
+	// Ack the first two messages.
+	err = ds.SendAcks(ctx, []driver.AckID{dms[0].AckID, dms[1].AckID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Ack them again; this should succeed even though we've acked them before.
+	// If providers return an error for this, drivers should drop them.
+	err = ds.SendAcks(ctx, []driver.AckID{dms[0].AckID, dms[1].AckID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Nack all 3 messages. This should also succeed, and the nack of the third
+	// message should take effect, so we should be able to fetch it again.
+	// Some providers may also re-send the first message.
```

Comment:
I don't really expect consistent behavior from an ack and a nack. In a significantly distributed system (like Google cloud pubsub), I believe they could happen in either order. So I'm not sure it's worth testing that combination.

Reply:
Agree, and this test doesn't test that (explicitly). I clarified this comment to be more clear that whether the first 2 messages are re-sent depends on provider semantics and timing. The 3rd message **must** be redelivered though.